// tcomb-form-templates-semantic/lib/list.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _getAlert = require('./getAlert');

var _getAlert2 = _interopRequireDefault(_getAlert);

var _renderFieldset = require('./renderFieldset');

var _renderFieldset2 = _interopRequireDefault(_renderFieldset);

function getButton(options) {
  return _react2['default'].createElement(
    'button',
    { key: options.key, className: 'ui basic button button-' + options.type, onClick: options.click },
    options.label
  );
}

function create() {
  var overrides = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

  function list(locals) {
    var children = [];

    if (locals.help) {
      children.push(list.renderHelp(locals));
    }

    if (locals.error && locals.hasError) {
      children.push(list.renderError(locals));
    }

    children = children.concat(locals.items.map(function (item) {
      return item.buttons.length === 0 ? list.renderRowWithoutButtons(item, locals) : list.renderRow(item, locals);
    }));

    if (locals.add) {
      children.push(list.renderAddButton(locals));
    }

    return list.renderFieldset(children, locals);
  }

  list.renderHelp = overrides.renderHelp || function renderHelp(locals) {
    return _getAlert2['default']('info', locals.help);
  };

  list.renderError = overrides.renderError || function renderError(locals) {
    return _getAlert2['default']('error', locals.error);
  };

  list.renderRowWithoutButtons = overrides.renderRowWithoutButtons || function renderRowWithoutButtons(item /* , locals*/) {
    return _react2['default'].createElement(
      'div',
      { key: item.key, className: 'six wide column' },
      _react2['default'].createElement(
        'div',
        { className: 'six wide column' },
        item.input
      )
    );
  };

  list.renderRowButton = overrides.renderRowButton || function renderRowButton(button) {
    return getButton({
      click: button.click,
      key: button.type,
      type: button.type,
      label: button.label
    });
  };

  list.renderButtonGroup = overrides.renderButtonGroup || function renderButtonGroup(buttons /* , locals*/) {
    return _react2['default'].createElement(
      'div',
      { className: 'ui basic buttons' },
      buttons.map(list.renderRowButton)
    );
  };

  list.renderRow = overrides.renderRow || function renderRow(row, locals) {
    return _react2['default'].createElement(
      'div',
      { key: row.key, className: 'ui grid' },
      _react2['default'].createElement(
        'div',
        { className: 'eight wide column' },
        row.input
      ),
      _react2['default'].createElement(
        'div',
        { className: 'four wide column' },
        list.renderButtonGroup(row.buttons, locals)
      )
    );
  };

  list.renderAddButton = overrides.renderAddButton || function renderAddButton(locals) {
    var style = {
      marginTop: '1em'
    };
    return _react2['default'].createElement(
      'div',
      { style: style },
      getButton(locals.add)
    );
  };

  list.renderFieldset = overrides.renderFieldset || _renderFieldset2['default'];

  list.clone = function clone() {
    var newOverrides = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    return create(_extends({}, overrides, newOverrides));
  };

  return list;
}

exports['default'] = create();
module.exports = exports['default'];
